test/cpp/api/inference_mode.cpp

```diff
@@ -407,3 +431,60 @@ TEST(InferenceModeTest, TestCreationMetaPropagation) {
   ASSERT_THROWS_WITH(c.add_(1),
     "A view was created in inference mode and is being modified inplace");
 }
+
+TEST(InferenceModeTest, TestInplaceCopyOnInferenceTensor) {
+  for (bool requires_grad: {true, false}) {
+    torch::Tensor s = torch::ones({1, 2, 3}).set_requires_grad(requires_grad);
+    torch::Tensor t;
+    {
+      InferenceMode guard;
+      t = torch::ones({1, 2, 3});
+      t.copy_(s);
+      ASSERT_TRUE(is_inference_tensor(t));
+      ASSERT_FALSE(t.requires_grad());
+    }
+
+    ASSERT_THROWS_WITH(t.copy_(s),
+      "Inplace update to inference tensor outside InferenceMode is not allowed");
+  }
+}
+
+TEST(InferenceModeTest, TestSetRequiresGradInNormalMode) {
+  torch::Tensor t;
+  {
+    InferenceMode guard;
+    t = torch::ones({1, 2, 3});
+  }
+  t.set_requires_grad(false);
+  ASSERT_THROWS_WITH(t.set_requires_grad(true),
+    "Setting requires_grad=True on inference tensor outside InferenceMode is not allowed.");
+}
+
+TEST(InferenceModeTest, TestAccessVersionCounter) {
+  torch::Tensor t;
+  {
+    InferenceMode guard;
+    t = torch::ones({1, 2, 3});
+    ASSERT_THROWS_WITH(t.unsafeGetTensorImpl()->version_counter().current_version(),
+      "Accessing version_counter of inference tensor is not allowed.");
+    t.unsafeGetTensorImpl()->bump_version();
+  }
+  ASSERT_THROWS_WITH(t.unsafeGetTensorImpl()->version_counter().current_version(),
+    "Accessing version_counter of inference tensor is not allowed.");
```

Comment:
Analogously, we could improve this error message too. In this case, it's easier to say what the problem is (rather than just say "not allowed"); just say inference tensors do not track version counters